product_state/models/product_template.py

```diff
@@ -1,18 +1,40 @@
 # Copyright 2017 ACSONE SA/NV (<http://acsone.eu>)
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
 
-from odoo import models, fields
+from odoo import api, models, fields
+
+
+class ProductState(models.Model):
+    _name = 'product.state'
+    _description = "Product State"
+    _order = 'name'
+
+    name = fields.Char('State Name', required=True)
+    code = fields.Char('State Code')
+    description = fields.Text(translate=True)
+    product_ids = fields.One2many(
+        'product.template',
+        'product_state_id',
+        string='State Products',
+    )
+    products_count = fields.Integer(
+        string='Number of products',
+        compute='_compute_products_count',
+    )
+
+    @api.multi
+    @api.depends('product_ids')
+    def _compute_products_count(self):
+        for state in self:
+            state.products_count = len(state.product_ids)
 
 
 class ProductTemplate(models.Model):
     _inherit = 'product.template'
 
-    state = fields.Selection(selection=[
-        ('draft', 'In Development'),
-        ('sellable', 'Normal'),
-        ('end', 'End of Lifecycle'),
-        ('obsolete', 'Obsolete')],
-        string='Status',
-        default='sellable',
-        index=True
+    product_state_id = fields.Many2one(
+        'product.state',
+        string='State',
+        help='Select a state for this product'
     )
+    state = fields.Char(related='product_state_id.code')
```

Comment:
why related='product_state_id.code' which is not required and does not show anywhere?
Maybe you want to write related='product_state_id.name'